crates/accelerate/src/commutation_checker.rs

```diff
@@ -0,0 +1,725 @@
+        op2_bound_qubits
+            .iter()
+            .for_each(|q| bq.add(py, &q, false).unwrap());
+        let qargs1 = op1_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+        let qargs2 = op2_bound_qubits
+            .iter()
+            .map(|q| bq.find(&q).unwrap().0)
+            .collect::<Vec<_>>();
+
+        let mut bc: BitData<Clbit> = BitData::new(py, "clbits".to_string());
+        let op1_bound_clbit = op1.instruction.clbits.bind(py);
+        let op2_bound_clbit = op2.instruction.clbits.bind(py);
+
+        op1_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        op2_bound_clbit
+            .iter()
+            .for_each(|c| bc.add(py, &c, false).unwrap());
+
+        let cargs1 = op1_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
+        let cargs2 = op2_bound_clbit
+            .iter()
+            .map(|c| bc.find(&c).unwrap().0)
+            .collect::<Vec<_>>();
+
+        Ok(self.commute_inner(
+            py,
+            &op1.instruction,
+            &qargs1,
+            &cargs1,
+            &op2.instruction,
+            &qargs2,
+            &cargs2,
+            max_num_qubits,
+        ))
+    }
+
+    #[pyo3(signature=(op1, qargs1, cargs1, op2, qargs2, cargs2, max_num_qubits=3))]
+    #[allow(clippy::too_many_arguments)]
+    fn commute(
+        &mut self,
+        py: Python,
+        op1: &Bound<PyAny>,
+        qargs1: Option<&Bound<PySequence>>,
+        cargs1: Option<&Bound<PySequence>>,
+        op2: &Bound<PyAny>,
+        qargs2: Option<&Bound<PySequence>>,
+        cargs2: Option<&Bound<PySequence>>,
+        max_num_qubits: u32,
+    ) -> PyResult<bool> {
+        let qargs1 =
+            qargs1.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let cargs1 =
+            cargs1.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let qargs2 =
+            qargs2.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        let cargs2 =
+            cargs2.map_or_else(|| Ok(PyTuple::empty_bound(py)), PySequenceMethods::to_tuple)?;
+        self.commute_nodes(
+            py,
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op1,
+                    Some(qargs1.into_any()),
+                    Some(cargs1.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
+            },
+            &DAGOpNode {
+                instruction: CircuitInstruction::py_new(
+                    op2,
+                    Some(qargs2.into_any()),
+                    Some(cargs2.into_any()),
+                )?,
+                sort_key: PyString::new_bound(py, "do_not_use").into_any().into(),
```

Comment:
You can use the `intern!` macro here to create a single pystring that gets reused to avoid the object creation overhead here.

```suggestion
                sort_key: intern!(py, "").into_any().into(),
            },
            &DAGOpNode {
                instruction: CircuitInstruction::py_new(
                    op2,
                    Some(qargs2.into_any()),
                    Some(cargs2.into_any()),
                )?,
                sort_key: intern!(py, "").into_any().into(),
```

Reply:
Since `sort_key` wants an object instead of a reference I only got this to work as
```rust
intern!(py, "").clone().into()
```
though this might undo the advantage of using `intern!`?

Reply:
`intern!` returns a `Py<PyString>`, which is a reference-counted smart pointer.  Calling `clone` on that increments the refcount and doesn't duplicate the underlying string.

(fwiw, `Py::clone_ref` is marginally more efficient than `Py::clone` when you have the `py` GIL token.)

Reply:
Nice thanks! This part of the code was actually removed in 454208cd9, which avoids the `CircuitInstruction` creation.